### PR TITLE
close / getUrl

### PR DESCRIPTION
**Pull request by pq**
Not sure if `CLOSE` really adds value and `getUrl` format is prime for refinement. Food for thought.

image

/cc @devoncarew

### CODE REVIEW

File: lib/src/channels/http.dart [hunk 1]
```diff
@@ -110,7 +117,7 @@ class LoggingHttpClient implements HttpClient {
 
   @override
   void close({bool force = false}) {
-    _todo('close');
+    _log.log('CLOSE${force ? " • (forced)" : ""}');
```
Comment (devoncarew) [inline lib/src/channels/http.dart]: I don't know if the `forced` is valuable to users. If we're printing `close` in the log, we should also saw what opened. From the HttpClient docs:

```
The HttpClient supports persistent connections and caches network connections to reuse them for multiple requests whenever possible. This means that network connections can be kept open for some time after a request has completed. Use HttpClient.close to force the HttpClient object to shut down and to close the idle network connections.
```

So the HttpClient could have 0 to n open network connections. Close might be useful to print if we could say when we open and close an underlying network connection (and not just a GET or POST request, several of which could run on a connection).

I guess I'd lean towards not logging anything for now, until as have a better way to measure what the underlying impl is doing.

File: lib/src/channels/http.dart [hunk 2]
```diff
@@ -135,9 +142,25 @@ class LoggingHttpClient implements HttpClient {
 
   @override
   Future<HttpClientRequest> getUrl(Uri url) {
-    // todo (pq): consider same logic as open (but w/ GET added)
-    _todo('getUrl');
-    return proxy.getUrl(url);
+    final int id = _nextRequestId++;
+    final String method = 'GET';
+
+    // #1 • GET • https://flutter.io url
```
Comment (devoncarew) [inline lib/src/channels/http.dart]: `#1 • GET $url` ?

File: lib/src/channels/http.dart [hunk 3]
```diff
@@ -135,9 +142,25 @@ class LoggingHttpClient implements HttpClient {
...
+    _log.log('#$id • $method • $url url');
+
+    Future<HttpClientRequest> request = proxy.getUrl(url);
+    return request.then((HttpClientRequest req) {
+      _log.log('#$id • $method • $url request ready');
```
Comment (devoncarew) [inline lib/src/channels/http.dart]: I'd be nice to determine which headers we're sending to the server (as well as the ones we get back, below).

File: lib/src/channels/http.dart [hunk 4]
```diff
@@ -44,8 +51,8 @@ class LoggingHttpClient implements HttpClient {
 
   @override
   set authenticateProxy(
-    Future<bool> Function(String host, int port, String scheme, String realm) f,
-  ) {
+      Future<bool> Function(String host, int port, String scheme, String realm) f,
+      ) {
```
[the remark this answers is not in the document]
Comment (pq) [reply lib/src/channels/http.dart]: Good eyes.  Fixed.

File: lib/src/channels/http.dart (hunk 1 above)
Comment (pq) [reply lib/src/channels/http.dart]: I was thinking the same thing.  Will remove.

File: lib/src/channels/http.dart (hunk 3 above)
Comment (pq) [reply lib/src/channels/http.dart]: Agreed.  TODO added.

Comment (pq) [reply lib/src/channels/http.dart]: Actually, easy enough to just add, so done.

**Comment by pq** [comment]
Thanks for feedback.

I opened an issue to track cleanup of remaining `http` log-worthy calls (#37) and created a new milestone to gather 1.0-related issues (https://github.com/pq/logs/milestone/1); feel free to chime in!

**Pull request merged**